tests/keygen-prov.c

```diff
@@ -0,0 +1,299 @@
+
+#include <libp11.h>
+
+static PKCS11_CTX *global_pkcs11_ctx = NULL;
+static PKCS11_SLOT *global_pkcs11_slots = NULL;
+static unsigned int global_pkcs11_slot_num;
+
+static int sign_verify_test(EVP_PKEY *priv, EVP_PKEY *pub)
+{
+	EVP_MD_CTX *mdctx = NULL;
+	int retval = 0;
+	char *msg = "libp11";
+	size_t slen;
+	unsigned char *sig = NULL;
+
+	if (!priv || !pub) {
+		printf("Where are the keys?\n");
+		return -1;
+	}
+	mdctx = EVP_MD_CTX_create();
+	if (!mdctx) {
+		display_openssl_errors();
+		retval = -2;
+		goto err;
+	}
+	if (1 != EVP_DigestSignInit(mdctx, NULL, EVP_sha256(), NULL, priv)) {
+		display_openssl_errors();
+		retval = -3;
+		goto err;
+	}
+	if (1 != EVP_DigestSignUpdate(mdctx, msg, strlen(msg))) {
+		display_openssl_errors();
+		retval = -4;
+		goto err;
+	}
+	if (1 != EVP_DigestSignFinal(mdctx, NULL, &slen)) {
+		display_openssl_errors();
+		retval = -5;
+		goto err;
+	}
+	if (!(sig = OPENSSL_malloc(sizeof(unsigned char) * (slen)))) {
+		display_openssl_errors();
+		retval = -6;
+		goto err;
+	}
+	if (1 != EVP_DigestSignFinal(mdctx, sig, &slen)) {
+		display_openssl_errors();
+		retval = -7;
+		printf("Sign fail\n");
+		goto err;
+	}
+	printf("Sign success\n");
+
+	if (1 != EVP_DigestVerifyInit(mdctx, NULL, EVP_sha256(), NULL, pub)) {
+		display_openssl_errors();
+		retval = -8;
+		goto err;
+	}
+	if (1 != EVP_DigestVerifyUpdate(mdctx, msg, strlen(msg))) {
+		display_openssl_errors();
+		retval = -9;
+		goto err;
+	}
+	if (1 == EVP_DigestVerifyFinal(mdctx, sig, slen)) {
+		printf("Verify success\n");
+		retval = 0;
+		goto err;
+	} else {
+		display_openssl_errors();
+		printf("Verify fail\n");
+		retval = -10;
+		goto err;
+	}
+
+err:
+	if(sig) OPENSSL_free(sig);
+	if(mdctx) EVP_MD_CTX_destroy(mdctx);
+	return retval;
+}
+
+static EVP_PKEY *setup_private(const char *token, const char *id, const char *obj,
+				const char *pin)
+{
+	char uri[200];
+
+	uri[0] = '\0';
+	strcat(uri, "pkcs11:token=");
+	strcat(uri, token);
+	strcat(uri, ";id=");
+	strcat(uri, id);
+	strcat(uri, ";object=");
+	strcat(uri, obj);
+	strcat(uri, ";type=private;pin-value=");
+	strcat(uri, pin);
+	printf("private uri: '%s'\n", uri);
+
+	return load_pkey(uri, NULL);
+}
+
+static EVP_PKEY *setup_public(const char *token, const char *id, const char *obj)
+{
+	char uri[200];
+
+	uri[0] = '\0';
+	strcat(uri, "pkcs11:token=");
+	strcat(uri, token);
+	strcat(uri, ";id=");
+	strcat(uri, id);
+	strcat(uri, ";object=");
+	strcat(uri, obj);
+	strcat(uri, ";type=public");
+	printf("public uri: '%s'\n", uri);
+
+	return load_pubkey(uri);
+}
+
+int main(int argc, char* argv[])
+{
+	int ret = EXIT_FAILURE;
+	PKCS11_SLOT *slot;
+
+	if (argc < 4) {
+		printf("Too few arguments\n");
+		printf("%s [TOKEN1] [KEY-LABEL] [PIN] [module]\n", argv[0]);
+		exit(ret);
+	}
+	const char *token_label = argv[1], *key_label = argv[2], *module = argv[4];
+	char *key_pass = argv[3];
+
+	/* Load pkcs11prov and default providers */
+	if (!providers_load()) {
```

Comment:
Unfortunately, the OpenSSL provider currently only supports retrieving objects from the token, meaning it cannot be used for key generation.
Please remove the key generation test for the provider.